src/Network/RateLimiting.hs

```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes #-}
+module Network.RateLimiting
+       ( -- * Types
+         RateLimiting(..)
+       , rlLift
+         -- * Rate-limiting strategies
+       , rateLimitingUnbounded
+       , rateLimitingFair
+       , rateLimitingBlocking
+       , rateLimitingDamping
+       ) where
+
+import           Control.Monad (void)
+import           Control.Monad.Trans.Class
+import qualified Data.Map.Strict as Map
+import           Data.Maybe (isJust, fromJust)
+import           Data.Time.Units (Microsecond)
+import           Mockable.Class
+import           Mockable.SharedAtomic
+import           Network.QDisc.Fair (fairQDisc)
+import           Network.Transport (EndPointAddress)
+import           Network.Transport.TCP (QDisc(..), simpleUnboundedQDisc)
+
+-- | A rate-limiting strategy.
+--
+-- This includes a 'QDisc' to govern how incoming requests are
+-- enqueued and dequeued, and also the possibility to block or delay
+-- requests from a peer depending on the number of in-flight bytes.
+data RateLimiting m =
+      NoRateLimiting !(forall t. IO (QDisc t))
+    | RateLimiting
+      { rlQDisc :: forall t. IO (QDisc t)
+      , rlLockByBytes
+            :: (Mockable SharedAtomic m)
+            => EndPointAddress
+            -> Int
+            -> m ()
+      , rlNewLock :: EndPointAddress -> m ()
+      , rlRemoveLock :: EndPointAddress -> m ()
+      }
+
+rlLift :: (Mockable SharedAtomic m, MonadTrans t) => RateLimiting m -> RateLimiting (t m)
+rlLift (NoRateLimiting qDisc) = NoRateLimiting qDisc
+rlLift RateLimiting{..} = RateLimiting
+    { rlQDisc = rlQDisc
+    , rlLockByBytes = \peer bytes -> lift $ rlLockByBytes peer bytes
+    , rlNewLock = lift . rlNewLock
+    , rlRemoveLock = lift . rlRemoveLock
+    }
+
+-- | The simplest rate-limiting procedure uses an unbounded queue, and
+-- performs no rate limiting at all.
+rateLimitingUnbounded :: Monad m => RateLimiting m
+rateLimitingUnbounded = NoRateLimiting simpleUnboundedQDisc
+
+-- | Ensures fariness in queueing requests from different peers, but
+-- does no rate-limiting.
+rateLimitingFair :: Monad m => RateLimiting m
+rateLimitingFair = NoRateLimiting (fairQDisc (const $ return Nothing))
```

Comment:
`rateLimitingUnbounded` and `rateLimitingFair` should be called `noRateLimitingUnbounded` and `noRateLimitingFair`.